### PR TITLE
feat(transport/cc): expose whether connection exited slow start as stat

### PR DESCRIPTION
**Pull request by omansfeld**
A first PR exposing the simplest stat discussed in #3112. Other stats like exit reason or approximations as to how close a heuristic was to the ideal exit point don't really apply as long as we only exit through loss/ecn and haven't implemented any heuristics.

But this is pretty useful already and would allow us to get some insight into how many connections actually exit slow start.

I hope with #3086 just merged we can also get this simple patch in quickly and then make a release so I can integrate both into Firefox glean metrics.

In addition to the unit tests below I also manually tested the changes using the `min_bandwidth` simulation with very small transfer amounts.

### CODE REVIEW

File: neqo-transport/src/cc/classic_cc.rs [hunk 1]
```diff
@@ -1411,7 +1420,7 @@ mod tests {
     #[test]
     fn spurious_congestion_event_detection() {
         let mut cc = ClassicCongestionControl::new(NewReno::default(), Pmtud::new(IP_ADDR, MTU));
-        let mut now = now();
```
Comment (omansfeld) [inline neqo-transport/src/cc/classic_cc.rs:1414]: small drive-by to make clippy happy, this slipped through #3086

File: neqo-transport/src/cc/classic_cc.rs [hunk 2]
```diff
@@ -345,8 +345,14 @@ impl<T: WindowAdjustment> CongestionControl for ClassicCongestionControl<T> {
...
+        // State needs to be saved before `on_congestion_event` is called where it changes to
+        // `State::RecoveryStart`.
+        let in_slow_start = self.state.in_slow_start();
         let congestion = self.on_congestion_event(last_lost_packet, now);
         if congestion {
+            if in_slow_start {
+                cc_stats.slow_start_exited = true;
+            }
             cc_stats.congestion_events_loss += 1;
         }
```
Comment (larseggert) [inline neqo-transport/src/cc/classic_cc.rs]:
```suggestion
        let was_in_slow_start = self.state.in_slow_start();
        let congestion = self.on_congestion_event(last_lost_packet, now);
        if congestion {
                cc_stats.slow_start_exited |= was_in_slow_start;
                cc_stats.congestion_events_loss += 1;
        }
```
You also have a very similar pattern below. Factor out into helper?

File: neqo-transport/src/cc/classic_cc.rs [hunk 3]
```diff
@@ -375,8 +381,14 @@ impl<T: WindowAdjustment> CongestionControl for ClassicCongestionControl<T> {
...
+        // State needs to be saved before `on_congestion_event` is called where it changes to
+        // `State::RecoveryStart`.
+        let in_slow_start = self.state.in_slow_start();
         let cwnd_reduced = self.on_congestion_event(largest_acked_pkt, now);
         if cwnd_reduced {
+            if in_slow_start {
+                cc_stats.slow_start_exited = true;
+            }
             cc_stats.congestion_events_ecn += 1;
         }
```
Comment (larseggert) [inline neqo-transport/src/cc/classic_cc.rs]:
```suggestion
        let was_in_slow_start = self.state.in_slow_start();
        let cwnd_reduced = self.on_congestion_event(largest_acked_pkt, now);
        if cwnd_reduced {
                cc_stats.slow_start_exited |= was_in_slow_start;
                cc_stats.congestion_events_ecn += 1;
        }
```
This is the "similar pattern" mentioned earlier.

File: neqo-transport/src/cc/classic_cc.rs
Comment (larseggert) [inline neqo-transport/src/cc/classic_cc.rs:379]:
```suggestion
        if !pkt.cc_outstanding() {
                return;
        }
```
Early return would remove indentation.

File: neqo-transport/src/cc/classic_cc.rs [hunk 4]
```diff
@@ -1527,4 +1536,38 @@ mod tests {
...
+        assert!(cc.state.in_slow_start());
+        assert!(!cc_stats.slow_start_exited);
+
+        let pkt1 = sent::make_packet(1, now, 1000);
+        cc.on_packet_sent(&pkt1, now);
+        cc.on_ecn_ce_received(&pkt1, now, &mut cc_stats);
+
+        assert!(!cc.state.in_slow_start());
+        assert!(cc_stats.slow_start_exited);
+    }
```
Comment (larseggert) [inline neqo-transport/src/cc/classic_cc.rs]: Can we extract the similarities between these tests into a helper?

File: neqo-transport/src/connection/tests/ecn.rs [hunk 5]
```diff
@@ -218,6 +218,7 @@ fn debug() {
   rx: 0 drop 0 dup 0 saved 0
   tx: 0 lost 0 lateack 0 ptoack 0 unackdrop 0
   cc: ce_loss 0 ce_ecn 0 ce_spurious 0
+  ss_exited: false
```
Comment (larseggert) [inline neqo-transport/src/connection/tests/ecn.rs]:
```suggestion
  ss_exit: false
```
Save two chars :-)

File: neqo-transport/src/stats.rs [hunk 6]
```diff
@@ -392,8 +390,9 @@ impl Debug for Stats {
             "  cc: ce_loss {} ce_ecn {} ce_spurious {}",
             self.cc.congestion_events_loss,
             self.cc.congestion_events_ecn,
-            self.cc.congestion_events_spurious
+            self.cc.congestion_events_spurious,
         )?;
+        writeln!(f, "  ss_exited: {}", self.cc.slow_start_exited)?;
```
Comment (larseggert) [inline neqo-transport/src/stats.rs]:
```suggestion
        writeln!(f, "  ss_exit: {}", self.cc.slow_start_exited)?;
```

File: neqo-transport/src/cc/classic_cc.rs (hunk 2 above)
Comment (omansfeld) [reply neqo-transport/src/cc/classic_cc.rs]: Helper not needed. It's possible to do this in `on_congestion_event` by passing in `ecn: bool` and thus avoid the duplication. We also need said bool if we ever want to do #3111.
